proposed/http-middleware/middleware-meta.md

```diff
@@ -136,59 +140,146 @@ only the request being passed to the middleware.
+The double pass approach is much newer but has been almost universally used by
 early adopters of HTTP Messages.
 
-5. `DelegateInterface`
+### 4.4 Chosen Approach
+
+Despite the nearly universal adoption of the double-pass approach there are
+significant issues regarding implementation.
+
+The most severe is that passing an empty response has no guarantees that the
+response is in a usable state. This is further exacerbated by the fact that a
+middleware may modify the response before passing it for further dispatching.
+
+Further compounding the problem is that there is no way to ensure that the
+response body has not been written to, which can lead to incomplete output or
+error responses being sent with cache headers attached. It is also possible
+to end up with [corrupted body content][rob-allen-filtering] when writing over
+existing body content if the new content is shorter than the original. The most
+effective way to resolve these issues is to always provide a fresh stream when
+modifying the body of a message.
+
+[rob-allen-filtering]: https://akrabat.com/filtering-the-psr-7-body-in-middleware/
+
+Some have argued that passing the response helps ensure dependency inversion.
+While it is true that it helps avoid depending on a specific implementation of
+HTTP messages, the problem can also be resolved by injecting factories into the
+middleware to create HTTP message objects, or by injecting empty message instances.
+With the creation of HTTP Factories in [PSR-17][psr17], a standard approach to
+handling dependency inversion is possible.
+
+[psr17]: https://github.com/php-fig/fig-standards/blob/master/proposed/http-factory/http-factory-meta.md
+
+A more subjective, but also important, concern is that existing double-pass
+middleware typically uses the `callable` type hint to refer to middleware.
+This makes strict typing impossible, as there is no assurance that the `callable`
+being passed implements a middleware signature, which reduces runtime safety.
+
+**Due to these significant issues the lambda approach has been choosen for this proposal.**
+
+5. Design Decisions
 -------------------
 
-The `$next` argument is a callable in most existing middleware systems. However using
-an interface allows to improve type safety and IDE support.
+### 5.1 Middleware Design
+
+The `MiddlewareInterface` defines a single method that accepts a server request
+and a delegate and must return a response. The middleware may:
+
+- Evolve the request before passing it to the delegate to execute the next
+  available middleware.
+- Evolve the response received from the delegate before returning it.
+- Create and return a response without passing it to the delegate, thereby
+  preventing any further middleware from executing.
+
+#### Why doesn't middleware use `__invoke`?
+
+Doing so would conflict with existing middleware that implements the double-pass
+approach and may want to implement the middleware interface.
```

Comment:
It is standard practice with FIG PSRs to maximize compatibility with existing systems, or at least attempt to not directly conflict. Type hinting `callable` is implicitly similar to using `__invoke`, so your argument about frameworks using `callable` doesn't make sense.